File: src/main/org/nlogo/window/ViewUpdatePanel.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.window;

import org.nlogo.api.I18N;
import org.nlogo.api.UpdateModeJ;

public strictfp class ViewUpdatePanel
    extends javax.swing.JPanel
    implements
    Events.LoadEndEventHandler {
  private final GUIWorkspace workspace;

  private final SpeedSliderPanel speedSlider;

  private final javax.swing.JComboBox<String> viewUpdates = new javax.swing.JComboBox<String>();

  private final WorldViewSettings settings;

  private int speed;

  public ViewUpdatePanel(GUIWorkspace workspace, final javax.swing.JCheckBox displaySwitch, boolean editable) {
    this.workspace = workspace;
    speedSlider = new SpeedSliderPanel(workspace, true);
    displaySwitch.addChangeListener
        (new javax.swing.event.ChangeListener() {
          public void stateChanged(javax.swing.event.ChangeEvent e) {
            boolean selected = displaySwitch.isSelected();
            if (selected != speedSlider.isEnabled()) {
              speedSlider.setEnabled(selected);
              if (selected) {
                speedSlider.setValue(speed);
              } else {
                speed = speedSlider.getValue();
                speedSlider.setValue(speedSlider.getMaximum());
              }
            }
          }
        });
    org.nlogo.awt.Fonts.adjustDefaultFont(displaySwitch);
    settings = workspace.viewWidget.settings();
    add(speedSlider);
    SettingsButton settingsButton = new SettingsButton();
    viewUpdates.addItem(I18N.guiJ().get("tabs.run.viewUpdates.dropdown.onticks"));
    viewUpdates.addItem(I18N.guiJ().get("tabs.run.viewUpdates.dropdown.continuous"));
    viewUpdates.setToolTipText(I18N.guiJ().get("tabs.run.viewUpdates.dropdown.tooltip"));
    // we don't want a settings button in the applet ev 2/28/06
    if (editable) {
      javax.swing.JPanel panel = new javax.swing.JPanel
          (new org.nlogo.awt.ColumnLayout
              (0,
                  java.awt.Component.CENTER_ALIGNMENT,
                  java.awt.Component.CENTER_ALIGNMENT));
      panel.add(displaySwitch);
      org.nlogo.awt.Fonts.adjustDefaultFont(viewUpdates);
      panel.setOpaque(false);
      panel.add(viewUpdates);
      add(panel);
      viewUpdates.addItemListener
          (new java.awt.event.ItemListener() {
            public void itemStateChanged(java.awt.event.ItemEvent e) {
              setMode((String) e.getItem());
            }
          });
      refreshSelection();
      add(new org.nlogo.swing.ToolBar.Separator());
      add(settingsButton);
    }
    setOpaque(true);
  }

  @Override
  public void setBackground(java.awt.Color color) {
    super.setBackground(color);
    if (speedSlider != null) {
      speedSlider.setBackground(color);
    }
    if (viewUpdates != null) {
      viewUpdates.setBackground(color);
    }
  }

  @Override
  public void addNotify() {
    super.addNotify();
    java.awt.Component[] comps = getComponents();
    for (int i = 0; i < comps.length; i++) {
      comps[i].setFocusable(false);
      org.nlogo.awt.Fonts.adjustDefaultFont(comps[i]);
      if (comps[i] instanceof javax.swing.AbstractButton &&
          // kinda kludgy but we don't want to have the text below
          // the checker in the checkbox in the Code tab ev 8/24/06
          !(comps[i] instanceof javax.swing.JCheckBox)) {
        ((javax.swing.AbstractButton) comps[i])
            .setVerticalTextPosition(javax.swing.SwingConstants.BOTTOM);
        ((javax.swing.AbstractButton) comps[i])
            .setHorizontalTextPosition(javax.swing.SwingConstants.CENTER);
      }
    }
  }

  private class SettingsButton
      extends javax.swing.JButton {
    public SettingsButton() {
      super(I18N.guiJ().get("tabs.run.settingsButton"));
      setFont
          (new java.awt.Font(org.nlogo.awt.Fonts.platformFont(),
              java.awt.Font.PLAIN, 10));
      setFocusable(false);
      setToolTipText(I18N.guiJ().get("tabs.run.settingsButton.tooltip"));
      addActionListener
          (new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent e) {
              new Events.EditWidgetEvent(settings)
                  .raise(SettingsButton.this);
            }
          });
    }
  }

  /// methods for keeping menu in sync with reality

  private void refreshSelection() {
    if(workspace.updateMode() == UpdateModeJ.TICK_BASED()) {
      viewUpdates.setSelectedIndex(0);
    }
    else if(workspace.updateMode() == UpdateModeJ.CONTINUOUS()) {
      viewUpdates.setSelectedIndex(1);
    }
    else {
      throw new IllegalStateException();
    }
  }

  private void setMode(String name) {
    if (name.equals(I18N.guiJ().get("tabs.run.viewUpdates.dropdown.onticks"))) {
      workspace.updateMode(UpdateModeJ.TICK_BASED());
    } else if (name.equals(I18N.guiJ().get("tabs.run.viewUpdates.dropdown.continuous"))) {
      workspace.updateMode(UpdateModeJ.CONTINUOUS());
    } else {
      throw new IllegalStateException();
    }
  }

  ///

  public void handle(Events.LoadEndEvent e) {
    refreshSelection();
    speedSlider.setValue((int) workspace.speedSliderPosition());
  }

}

File: src/main/org/nlogo/window/ViewUpdatePanel.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.window

import java.awt.{ Color, Component, Font }
import java.awt.event.{ ActionEvent, ActionListener, ItemEvent, ItemListener }
import javax.swing.{ AbstractButton, JButton, JCheckBox, JComboBox, JPanel, SwingConstants }
import javax.swing.event.{ ChangeEvent, ChangeListener }
import org.nlogo.api.{ I18N, UpdateMode }
import org.nlogo.awt.{ ColumnLayout, Fonts }
import org.nlogo.swing.ToolBar

class ViewUpdatePanel(workspace: GUIWorkspace, displaySwitch: JCheckBox, editable: Boolean)
    extends JPanel with Events.LoadEndEventHandler {
  implicit val i18nName = I18N.Prefix("tabs.run.viewUpdates.dropdown")
  private lazy val speedSlider = new SpeedSliderPanel(workspace, true)
  private lazy val viewUpdates = new JComboBox[String]
  private val settings = workspace.viewWidget.settings
  private val settingsButton = new SettingsButton
  private var speed = speedSlider.getValue
  
  displaySwitch.addChangeListener(new ChangeListener {
    def stateChanged(e: ChangeEvent) = {
      val selected = displaySwitch.isSelected
      if(selected != speedSlider.isEnabled) {
        speedSlider.setEnabled(selected)
        if(selected) {
          speedSlider.setValue(speed)
        } else {
          speed = speedSlider.getValue
          speedSlider.setValue(speedSlider.getMaximum)
        }
      }
    }
  })
  Fonts.adjustDefaultFont(displaySwitch)
  add(speedSlider)
  viewUpdates.addItem(I18N.gui("onticks"))
  viewUpdates.addItem(I18N.gui("continuous"))
  viewUpdates.setToolTipText(I18N.gui("tooltip"))
  // we don't want a settings button in the applet ev 2/28/06
  if(editable) {
    val panel = new JPanel(new ColumnLayout(0, Component.CENTER_ALIGNMENT, Component.CENTER_ALIGNMENT))
    panel.add(displaySwitch)
    Fonts.adjustDefaultFont(viewUpdates)
    panel.setOpaque(false)
    panel.add(viewUpdates)
    add(panel)
    viewUpdates.addItemListener(new ItemListener {
      def itemStateChanged(e: ItemEvent) = setMode(e.getItem.toString)
    })
    refreshSelection()
    add(new ToolBar.Separator)
    add(settingsButton)
  }
  setOpaque(true)

  override def setBackground(color: Color) = {
    super.setBackground(color)
    if(speedSlider != null)
      speedSlider.setBackground(color)
    if(viewUpdates != null)
      viewUpdates.setBackground(color)
  }

  override def addNotify() = {
    super.addNotify()
    val comps = getComponents
    comps.foreach { comp =>
      comp.setFocusable(false)
      Fonts.adjustDefaultFont(comp)
      comp match {
        // the `if` is kinda kludgy but we don't want to have the text below
        // the checker in the checkbox in the Code tab ev 8/24/06
        case button: AbstractButton if !button.isInstanceOf[JCheckBox] =>
          button.setVerticalTextPosition(SwingConstants.BOTTOM)
          button.setHorizontalTextPosition(SwingConstants.CENTER)
        case _ => {}
      }
    }
  }
  
  private class SettingsButton extends JButton(I18N.gui.get("tabs.run.settingsButton")) {
    setFont(new Font(Fonts.platformFont, Font.PLAIN, 10))
    setFocusable(false)
    setToolTipText(I18N.gui.get("tabs.run.settingsButton.tooltip"))
    addActionListener(new ActionListener {
      def actionPerformed(e: ActionEvent) = new Events.EditWidgetEvent(settings).raise(SettingsButton.this)
    })
  }

  /// methods for keeping menu in sync with reality

  private def refreshSelection() =
    if(workspace.updateMode() == UpdateMode.TickBased)
      viewUpdates.setSelectedIndex(0)
    else if(workspace.updateMode() == UpdateMode.Continuous)
      viewUpdates.setSelectedIndex(1)
    else
      throw new IllegalStateException

  private def setMode(name: String) =
    if(name==I18N.gui("onticks"))
      workspace.updateMode(UpdateMode.TickBased)
    else if (name==I18N.gui("continuous"))
      workspace.updateMode(UpdateMode.Continuous)
    else
      throw new IllegalStateException

  ///

  def handle(e: Events.LoadEndEvent) = {
    refreshSelection()
    speedSlider.setValue(workspace.speedSliderPosition.toInt)
  }

}
